### PR TITLE
tap: permit creating shallow taps

### PR DESCRIPTION
**Pull request by jonchang**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/master/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This will help speed up continuous integration runs when we really don't need a full clone but want to use the `brew tap` logic instead of e.g., `actions/checkout@v2`. This will also slightly speed up our self-hosted runner upgrades when installing casks.

See:

* https://discourse.brew.sh/t/force-shallow-tap/7367
* https://github.com/Homebrew/brew/pull/6991
* https://github.com/Homebrew/brew/pull/7004

### CODE REVIEW

File: Library/Homebrew/cmd/tap.rb [hunk 1]
```diff
@@ -51,11 +53,20 @@ def tap
     elsif args.no_named?
       puts Tap.names
     else
+      full_clone = if args.full?
+        true
+      elsif args.shallow?.nil?
+        !ENV["CI"]
+      else
+        !args.shallow?
```
Comment (MikeMcQuaid) [inline Library/Homebrew/cmd/tap.rb:61]:
```suggestion
      elsif !args.shallow?
        ENV["CI"].blank?
      else
        !args.shallow?
```

**Comment by jonchang** [comment]
Thanks for the feedback! Merging now so I can watch for problems on workflows today.

**Comment by jasonkarns** [comment]
Awesome, @jonchang! and thanks!

**Comment by jasonkarns** [comment]
@jonchang I'm seeing output that _seems_ to indicate that a shallow tap is being "unshallowed" during the tapping process.

Running on github Actions Ubuntu.
```
/home/runner/Homebrew/bin/brew tap --shallow homebrew/core
[/snipped output of fetching portable ruby]
Updating Homebrew...
==> Homebrew has enabled anonymous aggregate formulae and cask analytics.
Read the analytics documentation (and how to opt-out) here:
  https://docs.brew.sh/Analytics
No analytics have been recorded yet (or will be during this `brew` run).

==> Homebrew is run entirely by unpaid volunteers. Please consider donating:
  https://github.com/Homebrew/brew#donations
==> Tapping homebrew/core
Cloning into '/home/runner/Homebrew/Library/Taps/homebrew/homebrew-core'...
Tapped 2 commands and 5017 formulae (5,287 files, 13.9MB).
==> Unshallowing homebrew/core
```
workflow run: https://github.com/jasonkarns/setup-homebrew/runs/616411411?check_suite_focus=true

It's that final "unshallowing" that has me questioning what's going on. Thoughts? Should this be opened as a potential bug?

**Comment by jonchang** [comment]
Can you open a new issue with a reproducible example in e.g., a Docker container?

**Pull request merged**